### ISSUE TITLE
About VGGT scene scale computation on Scannet++ dataset

### ISSUE THREAD

**Issue by myc634** [facebookresearch/vggt#64]
Hi,

Sorry for bothering again, I try to reproduce some of the training process on scannet++ dataset, but I encountered some issue on scene scale normalization.

I try to directly follow the scene normalization process reported in the paper with the following code:
```
def normalize_cameras_batch_withpts3d(batches):
    intrinsics, extrinsics, pts3ds, valid_masks = batches["intrinsics"], batches["extrinsics"], batches["pts3d"], batches["valid_mask"] 
    bs = extrinsics.shape[0]
    normalized_cameras_c2ws, normalized_pts3ds, first_camera_w2cs, scene_scales = [], [], [], []

    for b in range(bs):
        camera_c2w, pts3d, valid_mask = extrinsics[b], pts3ds[b], valid_masks[b]
        # First, express all c2ws and pts3d in the first camera coordinate system
        first_camera_w2c = camera_c2w[0].inverse().unsqueeze(0).repeat(pts3d.shape[0], 1, 1)

        transformed_camera_c2w = torch.bmm(first_camera_w2c, camera_c2w)
        # Second, compute the Euclidean distance of all pts3d as scene scale
        pts3d_homo = torch.cat([pts3d, torch.ones_like(pts3d[:, :, :, :1])], dim=-1)
        transformed_pts3d = torch.bmm(first_camera_w2c, pts3d_homo.flatten(1, 2).transpose(1, 2)).transpose(1, 2)[..., :3]

        scene_scale = torch.norm(transformed_pts3d.flatten(0, 1)[valid_mask.flatten(0, 2).bool()], dim=-1).mean()

        # Third, normalize the camera c2w and pts3d
        transformed_camera_c2w[:, :3, 3] = transformed_camera_c2w[:, :3, 3] / scene_scale
        transformed_pts3d = transformed_pts3d / scene_scale

        normalized_cameras_c2ws.append(transformed_camera_c2w)
        normalized_pts3ds.append(transformed_pts3d)
        first_camera_w2cs.append(first_camera_w2c[0: 1])
        scene_scales.append(scene_scale)

    return torch.stack(normalized_cameras_c2ws), torch.stack(normalized_pts3ds), torch.stack(first_camera_w2cs), torch.stack(scene_scales)
```
I try to align the output by VGGT with the normalized ground truth point cloud, but seems their are some misalignment on the scene scale, it looks like this:

![Image](https://github.com/user-attachments/assets/e551d268-2bb7-40c1-9341-c6ad6d0bb32d)

it is clear that their are some misalignment issue on the right side of the wall and I can't figure out why.

Could you please give me some hint on how to reproduce the normalization method mentioned in the paper or point out what I am doing wrong?

It would be great if you can help me with that! Thanks for your patient and help!

**Comment by jianglh-WHU** [comment]
I encountered the same problem as well!

In the ScanNet++ dataset, the issue appears as described here: https://github.com/facebookresearch/vggt/issues/56#issuecomment-2755336719.
It's even more severe in the CO3D-v2 dataset. I'm not sure how to fix it. I pre-process co3dv2 using the scripts of dust3r.

Could you please provide some guidance on how to reproduce the normalization method mentioned in the paper?

<img width="600" alt="Image" src="https://github.com/user-attachments/assets/0407eb22-7ec6-4bd5-9633-bc7e7a963e5a" />

**Comment by jytime** [comment]
Hi @jianglh-WHU  @myc634 ,

Here is the dirty code we used for normalization

```

def normalize_camera_extrinsics_and_points_batch(
    extrinsics,
    cam_points=None,
    world_points=None,
    depths=None,
    scale_by_points=True,
    point_masks=None,
    mode="mean",
    seq_name=None,
):
    # Note this assumes we use cpu
    # extrinsics: (B, S, 3, 4)
    # world_points: (B, S, H, W, 3) or (*,3)
    # cam_points: same shape as world_points or something consistent
    # point_masks: (B, S, H, W) boolean mask if provided

    check_valid_tensor(extrinsics, "extrinsics")
    check_valid_tensor(cam_points, "cam_points")
    check_valid_tensor(world_points, "world_points")
    check_valid_tensor(depths, "depths")


    B, S, _, _ = extrinsics.shape
    device = extrinsics.device
    dtype = extrinsics.dtype


    # Convert extrinsics to homogeneous form: (B, N,4,4)
    extrinsics_homog = torch.cat(
        [
            extrinsics,
            torch.zeros((B, S, 1, 4), device=device),
        ],
        dim=-2,
    )
    extrinsics_homog[:, :, -1, -1] = 1.0

    # first_cam_extrinsic_inv, the inverse of the first camera's extrinsic matrix
    # which can be also viewed as the cam_to_world extrinsic matrix
    first_cam_extrinsic_inv = closed_form_inverse_se3(extrinsics_homog[:, 0])
    # new_extrinsics = torch.matmul(extrinsics_homog, first_cam_extrinsic_inv)
    new_extrinsics = torch.matmul(extrinsics_homog, first_cam_extrinsic_inv.unsqueeze(1))  # (B,N,4,4)

    # Force the first camera to be the identity
    # Do we really need this? Close it now
    # identity_4x4 = torch.eye(4, device=device, dtype=dtype)
    # new_extrinsics[:, 0] = identity_4x4


    if world_points is not None:
        # since we are transforming the world points to the first camera's coordinate system
        # we directly use the cam_from_world extrinsic matrix of the first camera
        # instead of using the inverse of the first camera's extrinsic matrix
        R = extrinsics[:, 0, :3, :3]
        t = extrinsics[:, 0, :3, 3]
        new_world_points = (world_points @ R.transpose(-1, -2).unsqueeze(1).unsqueeze(2)) + t.unsqueeze(1).unsqueeze(2).unsqueeze(3)
    else:
        new_world_points = None


    if scale_by_points:
        new_cam_points = cam_points.clone()
        new_depths = depths.clone()

        dist = new_world_points.norm(dim=-1)
        dist_sum = (dist * point_masks).sum(dim=[1,2,3])
        valid_count = point_masks.sum(dim=[1,2,3])
        avg_scale = (dist_sum / (valid_count + 1e-3)).clamp(min=1e-3, max=1e3)


        new_world_points = new_world_points / avg_scale.view(-1, 1, 1, 1, 1)
        new_extrinsics[:, :, :3, 3] = new_extrinsics[:, :, :3, 3] / avg_scale.view(-1, 1, 1)
        if depths is not None:
            new_depths = new_depths / avg_scale.view(-1, 1, 1, 1)
        if cam_points is not None:
            new_cam_points = new_cam_points / avg_scale.view(-1, 1, 1, 1, 1)
    else:
        return new_extrinsics[:, :, :3], cam_points, new_world_points, depths


```

Please note that it is not expected the model outputs should exactly match the GT scale. During training, even the same scene, the scale is slightly different using a different subset of images, e.g., images [0,1,3,10,11] will have a different GT scale to images [0,2,6,10,12]. We do not expect or hope the model to fit to such a scale. So it is correct if the predicted point map does not exactly match the GT normalised point map.

**Comment by myc634** [comment]
Hi,

Many thanks for your quick reply, it helps a lot!

I carefully read the code and I think the normalization code we use are actually equal. The thing is: let's say we use the depth and pose from frame [0, 1, 3, 5, 7] to compute the scale and do normalization, and we also send the rgb image from frame [0, 1, 3, 5, 7] into the vggt model as input, in my understanding, theoretically, we should have at least the same scale of the output point map, rather than those point map with some scale error. If the pointmap has some misalignment on scale, even though vggt model gives a super great output on geometry, I guess the mse loss of point regression might be pretty high.

Please correct me if my understanding is wrong! Thanks again for your quick and responsible reply.

**Comment by jianglh-WHU** [comment]
Thanks for your quick reply!

But from my understanding, I think if there is a misalignment between the GT normalized point maps and the VGGT predicted point maps, both the depth loss and point map loss will be very high, or may even fail to converge. Please correct me if I'm wrong!

**Comment by jytime** [comment]
Hi the release model reached a depth loss of around 0.01-0.02, which it looks good to me. During training, the loss curve is quite stable. 

You could easily see similar observations using the training code of Dust3R, e.g., just removing their normalisation over the predictions.

**Comment by myc634** [comment]
Hi!

Thanks for your quick reply, I have checked the mse loss of the point prediction, it do seems around this range. but I also check the dust3r with GA, seems they don't suffers from this kind of misalignment on scale. Looks like this is a potential improvement of VGGT :)

Thanks for your reply again and I truly appreciate your contribution to the community, VGGT is a great work!

**Comment by lifuguan** [comment]
Hi!

Thanks for the reply. I've checked the normalization code and tried to reproduce the depth loss (here we only normalize the GT depth):
```python
'''Training Loop'''
with torch.amp.autocast('cuda', enabled=False):
    new_extrinsics, _, new_world_points, depths = normalize_camera_extrinsics_and_points_batch(
                                                    extrinsics = batch['extrinsic'], 
                                                    cam_points = None, 
                                                    world_points = batch['world_points'], 
                                                    depths = batch['depth'],
                                                    point_masks = batch['valid_mask'])
    batch['extrinsic'] = new_extrinsics
    batch['world_points'] = new_world_points
    batch['depth'] = depths
    
    final_loss, loss_details = train_criterion(batch, predictions) if train_criterion is not None else None
loss_value = float(final_loss)

'''Depth Loss'''
class RegrDepth (Criterion, MultiLoss):
    """ Ensure that all 3D points are correct.
        Asymmetric loss: view1 is supposed to be the anchor.

        P1 = RT1 @ D1, point clouds at world coord_frame
        P2 = RT2 @ D2
        loss1 = (I @ pred_D1) - (RT1^-1 @ RT1 @ D1), I is identical matrix
        loss2 = (RT21 @ pred_D2) - (RT1^-1 @ P2)
              = (RT21 @ pred_D2) - (RT1^-1 @ RT2 @ D2)
    """

    def __init__(self, criterion, norm_mode='avg_dis', gt_scale=False):
        super().__init__(criterion)
        self.norm_mode = norm_mode
        self.gt_scale = gt_scale

    def get_all_pts3d(self, gt, pred, dist_clip=None):
        pseudo_focal = torch.cat([torch.full(gt['world_points'][..., 0].shape, gt['intrinsic'][0,0,0,0].item()).unsqueeze(1),
                                  torch.full(gt['world_points'][..., 0].shape, gt['intrinsic'][0,0,1,1].item()).unsqueeze(1)], dim=1).to(gt['depth'].device)

        B, N, H, W, _ = gt['depth'].shape  # 获取原始形状
        focal_reshaped = pseudo_focal.reshape(B * N, 2, H, W)

        gt_depth_reshaped = gt['depth'].reshape(B * N, H, W)
        gt_depth_pts = depthmap_to_pts3d(gt_depth_reshaped, focal_reshaped)

        pred_depth_reshaped = pred['depth'].reshape(B * N, H, W)
        pred_depth_pts = depthmap_to_pts3d(pred_depth_reshaped, focal_reshaped)

        # 如果没有valid 则设置为全1
        if 'valid_mask' not in gt:
            valid = torch.ones_like(gt['depth'][..., 0], dtype=torch.bool)
        else:
            valid = gt['valid_mask'].clone()
        valid = valid.reshape(B * N, H, W)

        if dist_clip is not None:
            # points that are too far-away == invalid
            dis = gt_depth_pts.norm(dim=-1)  # (B, H, W)
            valid = valid & (dis <= dist_clip)

        # normalize 3d points
        # if self.norm_mode:
        #     pred_depth_pts = normalize_pointcloud(pred_depth_pts, self.norm_mode, valid)
        # if self.norm_mode and not self.gt_scale:
        #     gt_depth_pts = normalize_pointcloud(gt_depth_pts, self.norm_mode, valid)

        gt_depth_pts = gt_depth_pts.reshape(B, N, H, W, 3)
        pred_depth_pts = pred_depth_pts.reshape(B, N, H, W, 3)
        valid = valid.reshape(B, N, H, W)

        return gt_depth_pts, pred_depth_pts, valid, {}

    def compute_loss(self, gt, pred, **kw):
        gt_pts, pred_pts, mask, monitoring = \
            self.get_all_pts3d(gt, pred, **kw)
        # loss on img1 side
        l1 = self.criterion(pred_pts[mask], gt_pts[mask])
        self_name = type(self).__name__
        details = {self_name: float(l1.mean())}
        return (l1, mask), (details | monitoring)

```
However, the depth loss seems very large (i.e., 9.0-10.0). Is there anything wrong?

**Comment by myc634** [comment]
Seems your projected pts3d are not in the same coords, did you do visualization checking on your code?

**Comment by myc634** [comment]
Looks confused to me, are you trying to compute depth loss or pts loss? If you are trying to compute depth loss, seems you are using gt depth to supervise unprojected point map.

**Comment by lifuguan** [comment]
Hi, here I follow DUSt3R to unproject both GT and predicted depth into point clouds and perform supervision. If not, would you like to share your depth supervision code?

**Comment by lifuguan** [comment]
> Looks confused to me, are you trying to compute depth loss or pts loss? If you are trying to compute depth loss, seems you are using gt depth to supervise unprojected point map.

Meanwhile, I noticed a significant distribution difference between the predicted depth and the ground truth depth, which might be the reason of the unusual loss value?

![Image](https://github.com/user-attachments/assets/2f282621-1f21-4dab-a601-69d093eda1a9)

**Comment by myc634** [comment]
Hmmm, the code you post here seems like the depth is still depth map rather than unprojected point map

```
new_extrinsics, _, new_world_points, depths = normalize_camera_extrinsics_and_points_batch(
                                                    extrinsics = batch['extrinsic'], 
                                                    cam_points = None, 
                                                    world_points = batch['world_points'], 
                                                    depths = batch['depth'],
                                                    point_masks = batch['valid_mask'])
    batch['extrinsic'] = new_extrinsics
    batch['world_points'] = new_world_points
    batch['depth'] = depths
```

**Comment by lifuguan** [comment]
> Hmmm, the code you post here seems like the depth is still depth map rather than unprojected point map
> 
> ```
> new_extrinsics, _, new_world_points, depths = normalize_camera_extrinsics_and_points_batch(
>                                                     extrinsics = batch['extrinsic'], 
>                                                     cam_points = None, 
>                                                     world_points = batch['world_points'], 
>                                                     depths = batch['depth'],
>                                                     point_masks = batch['valid_mask'])
>     batch['extrinsic'] = new_extrinsics
>     batch['world_points'] = new_world_points
>     batch['depth'] = depths
> ```

Yea. The unproject process are conducted in the `RegerDepth` class:
```python
        gt_depth_reshaped = gt['depth'].reshape(B * N, H, W)
        gt_depth_pts = depthmap_to_pts3d(gt_depth_reshaped, focal_reshaped)

        pred_depth_reshaped = pred['depth'].reshape(B * N, H, W)
        pred_depth_pts = depthmap_to_pts3d(pred_depth_reshaped, focal_reshaped)
```

**Comment by myc634** [comment]
The loss function I use is just simply F.mes_loss between the prediction point map and the normalized gt point map and the number is pretty close to the number that author posted here :)

**Comment by myc634** [comment]
I see the code here, if everything correct, directly apply F.mes_loss between `gt_depth_pts` and `pred_depth_pts` should get a good result, and don't forget the `validmask`

**Comment by lifuguan** [comment]
Got it. Thanks for the reply again! And have you tested on TartainAir dataset? I wonder if this could be an issue with the TartanAir dataset.

**Comment by myc634** [comment]
Unfortunately no, my result on scannetpp looks fine, just reuse the code for dataset in dust3r codebase will works

**Comment by lifuguan** [comment]
Ooops, I just tested the MSE loss, the loss value is still very strange😢

<img width="493" alt="Image" src="https://github.com/user-attachments/assets/4a393be2-2b0a-4508-ac7b-1cc5612f1bd8" />

**Comment by myc634** [comment]
Seems you forget the `valid mask` and I strongly advise to do the visualization check. As I can only see a part of your code, I highly suspect that the `gt['depth']` and `pred['depth']`  are actually not the same thing. From your code posted here, I recommend to set your breakpoint inside the function  `get_all_pts3d` and compute the mse loss between `gt_depth_pts` and `pred_depth_pts`

**Comment by lifuguan** [comment]
The `valid mask` from TartanAir is set to all `True` since it is a synthesis dataset.

**Comment by myc634** [comment]
I still recommend to do valid masking, as TartanAir is an outdoor dataset, their existing some sky region should be masked.

**Comment by njfan** [comment]
I don't want to normalize my GT dataset, how can I restore the obtained point cloud to the GT world coordinate system?

**Comment by Shexiaox** [comment]
> I don't want to normalize my GT dataset, how can I restore the obtained point cloud to the GT world coordinate system?

Hello, I've encountered the same issue. Have you found a solution?
